source/com.microsoft.tfs.client.common.ui/src/com/microsoft/tfs/client/common/ui/controls/connect/ServerTypeSelectControl.java

```diff
@@ -114,33 +157,112 @@ public void widgetSelected(final SelectionEvent e) {
+        serverCombo.addModifyListener(new ModifyListener() {
             @Override
-            public void selectionChanged(final SelectionChangedEvent event) {
+            public void modifyText(final ModifyEvent e) {
                 if (!ignoreServerChangeEvents) {
-                    setServer(serverControl.getServerURI());
+                    setServer(getServerFromCombo(), false);
                 }
             }
         });
 
-        // fire the buttons changed event so that the initial state is updated
-        onButtonsChanged();
+        final Button serversButton =
+            SWTUtil.createButton(container, Messages.getString("ServerSelectControl.ServersButtonText")); //$NON-NLS-1$
+        serversButton.addSelectionListener(new SelectionAdapter() {
+            @Override
+            public void widgetSelected(final SelectionEvent e) {
+                onServersButtonSelected();
+            }
+        });
+
+        GridDataBuilder.newInstance().hSpan(parentLayout).hIndent(getHorizontalSpacing() * 4).hFill().hGrab().vIndent(
+            getVerticalSpacing()).applyTo(container);
+
+        return container;
+    }
+
+    private void populateServersCombo() {
+        final ServerList serverList =
+            ServerListManagerFactory.getServerListProvider(DefaultPersistenceStoreProvider.INSTANCE).getServerList();
+
+        int i = 0;
+        final Set<ServerListConfigurationEntry> servers = serverList.getServers();
+        final String[] list = new String[servers.size()];
+        for (final ServerListConfigurationEntry entry : servers) {
+            list[i++] = entry.getURI().toString();
+        }
+        Arrays.sort(list);
+        serverCombo.setItems(list);
+        proposalProvider.setProposals(list);
+
+        if (list.length == 1) {
+            serverCombo.select(0);
+        }
+    }
+
+    protected void onServersButtonSelected() {
+        final URI lastSelectedServer = getServerFromCombo();
+        final ServerList serverList =
+            ServerListManagerFactory.getServerListProvider(DefaultPersistenceStoreProvider.INSTANCE).getServerList();
+
+        final ServerList serverListCopy = new ServerList();
+        serverListCopy.addAll(serverList.getServers());
+
+        final ServerListDialog dialog = new ServerListDialog(getShell(), serverListCopy);
+        dialog.open();
+
+        /*
+         * Note: there is no OK/Cancel on ServerListDialog: all exit paths
+         * should continue.
+         */
+
+        ServerListManagerFactory.getServerListProvider(DefaultPersistenceStoreProvider.INSTANCE).setServerList(
+            serverListCopy);
+
+        populateServersCombo();
+
+        final ServerListConfigurationEntry lastAddedServerListEntry = dialog.getLastAddedServerListEntry();
+
+        if (lastAddedServerListEntry != null) {
+            setServerOnCombo(lastAddedServerListEntry.getURI());
+        } else {
+            setServerOnCombo(lastSelectedServer);

```

Comment:
1. Shouldn't we use the server selected in the dialog in this case?
2. lastSelectedServer might be deleted in the dialog.